### ISSUE TITLE
DataLoader may be too slow

### ISSUE THREAD

**Issue by nekosan314** [ando-hub/MSA_Pretrain#3]
Hi Ando, I have some difficulty.

When I run ```train.py``` to train this model, it takes about 1 hour to run 1 epoch.
I think it may be  too late, but does it work as expected?

I found that the cause of delay may be DataLoader. Most of the time is used to get iteration. And I also found parameters such as num_workers and pin_memory on DataLoader is effective to speed up DataLoader.

Thus I set num_workers=8 on ```train.sh```(8 is my machine's upper limitation), ```pin_memory=True```on DataLoader, but they had no effect to speed up.

Do you have any solution?

**Comment by ando-hub** [comment]
I'm afraid you are running on CPU or with hyperparameters that I haven't tried. Please tell me your environment and the parameters you tried.
Note that in my environment (Intel(R) Xeon(R) Platinum 8260 CPU @ 2.40GHz 48core, NVIDIA RTX A6000), num_workers=8 and on GPU, it took 7/15min to process 1 epoch by `layerbest`/`layerall` config files.

**Comment by nekosan314** [comment]
My environment is as follows.

- Xeon E3-1246 v3 3.50GHz(4C/8T)
- GTX 1080 Ti(11GB)

And hyperparameters is as follows.

- enc1sap4dec1_gate_dec2.yaml
- mb16ep50adam0.0001schedspecaugseed0.yaml
- layer23.yaml

I ran ```nvidia-smi``` and noticed that GPU memory was used but GPU-util was pretty low(almost always 0%, but sometimes ). I also found that CUDA 0 memory is used when I set gpuid=1 on ```train.sh```. 
I think these things indicate that I'm runnning onCPU as you said, and this may be caused by my environment. Thus I will review my environment.

**Comment by ando-hub** [comment]
Thank you for your information. Your env and setup are almost the same as mine, so it should work well. The following tips may solve the problem;
1. The environmental variable `CUDA_VISIBLE_DEVICES` may be mismatched from what you set in `train.sh`. Please try to set the appropriate `CUDA_VISIBLE_DEVICES`, or replace `train.py` L254 as: `os.environ["CUDA_VISIBLE_DEVICES"] = str(args.gpu); device = 'cuda'`.
2. The dataloader access multiple .npy files, which will increase disk I/O and decrease process speed. I recommend copying the feature dirs to your local disk, e.g., `/tmp` or `/var/tmp`, if you use storage over networks.

**Comment by nekosan314** [comment]
Thank you for your advices, but my situation didn't improve.

1. I change ```os.environ["CUDA_VISIBLE_DEVICES"] = str(args.gpu); device = 'cuda' ```  in ```train.py``` L254. But GPU-util is still almost always 0% and sometimes about 10% or so. 
(On the other hand GPU's memory which was appointed by ```os.environ["CUDA_VISIBLE_DEVICES" ``` is used.) 
I think such a low GPU-util is abnormal, isn't it?
3. In fact I have already copied feature data to ```/tmp```, and this made run-time fast. But I still have to wait about 1 hour to run 1  epoch.

**Comment by ando-hub** [comment]
> I think such a low GPU-util is abnormal, isn't it?
Actually, it is the same in my environments; GPU util is from 0 to 10% during training. I used the small batchsize because it showed better in prediction accuracy.

Here is my training log on GPU,
```
[2023-01-10 19:28:13,200] INFO: load dataset ...
[2023-01-10 19:28:58,478] INFO: start training ...
[2023-01-10 19:37:51,481] INFO: Ep: 1/50 [acc]  trn=0.737, dev=0.812 tst=0.816 (proc:8m53s, rest:7h15m17s)
[2023-01-10 19:42:19,712] INFO: Ep: 2/50 [acc]  trn=0.816, dev=0.830 tst=0.842 (proc:13m21s, rest:5h20m29s)
[2023-01-10 19:46:06,601] INFO: Ep: 3/50 [acc]  trn=0.837, dev=0.833 tst=0.850 (proc:17m8s, rest:4h28m27s)
```
and on CPU (sometimes run faster than GPU when I use small batchsize)
```
[2023-01-10 19:52:32,612] INFO: load dataset ...
[2023-01-10 19:52:42,906] INFO: start training ...
[2023-01-10 19:56:13,463] INFO: Ep: 1/50 [acc]  trn=0.736, dev=0.812 tst=0.818 (proc:3m30s, rest:2h51m57s)
[2023-01-10 19:59:55,148] INFO: Ep: 2/50 [acc]  trn=0.816, dev=0.840 tst=0.845 (proc:7m12s, rest:2h52m53s)
```
Please report your training log on GPU and CPU if possible.

Now I'm participating in SLT @ Doha, and difficult to make time for fixing issues. I'll check my code after going back to my country. I'm sorry for the inconvenience.

**Comment by nekosan314** [comment]
Thank you for responding despite your busy schedule.

Here is my train log on GPU
```
[2023-01-10 21:55:05,574] INFO: start training ...
[2023-01-10 22:54:46,689] INFO: Ep: 1 save model ...
[2023-01-10 22:54:47,088] INFO: Ep: 1/50 [loss] trn=0.411, dev=0.400 tst=0.390
[2023-01-10 22:54:47,089] INFO: Ep: 1/50 [acc]  trn=0.583, dev=0.618 tst=0.629 (proc:59m41s, rest:48h44m54s)
[2023-01-10 23:56:47,376] INFO: Ep: 2 save model ...
[2023-01-10 23:56:48,163] INFO: Ep: 2/50 [loss] trn=0.377, dev=0.376 tst=0.364
[2023-01-10 23:56:48,164] INFO: Ep: 2/50 [acc]  trn=0.650, dev=0.673 tst=0.686 (proc:2h1m42s, rest:48h41m2s)
```
and on CPU
```
[2023-01-11 01:09:04,551] INFO: Ep: 1/50 [loss] trn=0.412, dev=0.400 tst=0.391
[2023-01-11 01:09:04,552] INFO: Ep: 1/50 [acc]  trn=0.584, dev=0.616 tst=0.628 (proc:1h1m39s, rest:50h21m2s)
[2023-01-11 02:11:04,871] INFO: Ep: 2 save model ...
[2023-01-11 02:11:05,639] INFO: Ep: 2/50 [loss] trn=0.377, dev=0.380 tst=0.367
[2023-01-11 02:11:05,640] INFO: Ep: 2/50 [acc]  trn=0.653, dev=0.696 tst=0.710 (proc:2h3m40s, rest:49h28m7s)
```

I found most of the time was spent between L110 and L111 on ```train.py```, so I thought this problem is caused by dataloader of pytorch. 

I will continue to look for problems until you return. Looking forward to hearing from you.

**Comment by nekosan314** [comment]
I was able to solve the problem by changing the environment!
I changed my environment as follows.
- Core i9-9820X 3.3GHz(10C/20T)
- RTX 2080 Ti(11GB)

I don't know why the problem had been solved , but I think CPU specs might be problem.
Thank you for your cooperation in spite of your busy schedule!
